core/theme/ThemeManager.php

```diff
@@ -0,0 +1,227 @@
+<?php
+
+namespace luya\theme;
+
+use luya\base\PackageConfig;
+use luya\Exception;
+use luya\helpers\Json;
+use Yii;
+use yii\base\InvalidArgumentException;
+use yii\base\InvalidConfigException;
+
+/**
+ * Core theme manager for LUYA.
+ *
+ * This component manage available themes via file system and the actual display themes.
+ *
+ * @author Mateusz Szymański Teamwant <zzixxus@gmail.com>
+ * @author Bennet Klarhölter <boehsermoe@me.com>
+ * @since  1.1.0
+ */
+class ThemeManager extends \yii\base\Component
+{
+    const APP_THEMES_BLANK = '@app/themes/blank';
+    
+    /**
+     * Name of the theme which should be activated on setup.
+     *
+     * @var string
+     */
+    public $activeThemeName;
+    
+    /**
+     * @var ThemeConfig[]
+     */
+    private $_themes = [];
+    
+    /**
+     * Read the theme.json and create a new \luya\theme\ThemeConfig for the given base path.
+     *
+     * @param string $basePath
+     *
+     * @return ThemeConfig
+     * @throws Exception
+     * @throws InvalidConfigException
+     */
+    protected static function loadThemeConfig(string $basePath): ThemeConfig
+    {
+        if (strpos($basePath, '@') === 0) {
+            $dir = Yii::getAlias($basePath);
+        } elseif (strpos($basePath, '/') !== 0) {
+            $dir = $basePath = Yii::$app->basePath . DIRECTORY_SEPARATOR . $basePath;
+        }
+        
+        if (!is_dir($dir) || !is_readable($dir)) {
+            throw new Exception('Theme directory not exists or readable: ' . $dir);
+        }
+        
+        $themeFile = $dir . '/theme.json';
+        if (file_exists($themeFile)) {
+            $config = Json::decode(file_get_contents($themeFile)) ?: [];
+        } else {
+            $config = [];
+        }
+        
+        $themeConfig = new ThemeConfig($basePath, $config);
+        
+        return $themeConfig;
+    }
+    
+    /**
+     * Setup active theme
+     */
+    final public function setup()
+    {
+        if ($this->getActiveTheme() instanceof Theme) {
+            // Active theme already loaded
+            return;
+        }
+        
+        $basePath = $this->getActiveThemeBasePath();
+        
+        try {
+            $this->beforeSetup($basePath);
+            
+            $themeConfig = $this->getThemeByBasePath($basePath);
+            $theme = new Theme($themeConfig);
+            $this->activate($theme);
+        } catch (InvalidArgumentException $ex) {
+            Yii::error($ex->getMessage(), 'luya-theme');
```

Comment:
setting up a theme will then turn into a silent failure which is not trackable aslong as debug toolbar is disabled? Is this expected? If a setup a theme, and its failing it should throw an exception i would say. So whats the reason behind?

Reply:
In the bootstrap section there is always: ` $app->themeManager->setup();` but the setup() does not take care whether there is no theme to setup, instead it creates the above mentioned silent error.

+ can we determine whether themes are set or not? if yes we should to this check on the first line of setup.

Reply:
If think about it make no sense to catch this exception. When something goes wrong with the theme setup there should an error occur.